File: backend/src/main/java/com/noomit/backend/reception/domain/ServiceRequest.java
package com.noomit.backend.reception.domain;

import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalTime;

public record ServiceRequest(
        Long id,
        String requestNumber,
        long customerId,
        Long productId,
        Long selectedSubCategoryId,
        String selectedModelName,
        long receptionistId,
        String symptom,
        String remarks,
        int baseFee,
        Long reservedSlotId,
        Long technicianId,
        LocalDate visitDate,
        LocalTime visitStartTime,
        LocalTime visitEndTime,
        ServiceRequestStatus status,
        Instant requestedAt,
        Instant assignedAt,
        Instant cancelledAt,
        String cancelReason,
        long version,
        Instant createdAt,
        Instant updatedAt) {

    public static ServiceRequest create(String requestNumber, long customerId, Long productId,
                                         Long selectedSubCategoryId, String selectedModelName, long receptionistId,
                                         String symptom, String remarks, int baseFee, Instant requestedAt) {
        return new ServiceRequest(
                null, requestNumber, customerId, productId, selectedSubCategoryId, selectedModelName,
                receptionistId, symptom, remarks, baseFee,
                null, null, null, null, null,
                ServiceRequestStatus.RECEIVED,
                requestedAt, null, null, null,
                0L, requestedAt, requestedAt
        );
    }

    public boolean canAssign() {
        return status == ServiceRequestStatus.RECEIVED;
    }

    public boolean canReassign() {
        return status == ServiceRequestStatus.ASSIGNED;
    }

    public boolean canCancel() {
        return status == ServiceRequestStatus.RECEIVED || status == ServiceRequestStatus.ASSIGNED;
    }

    public boolean canEdit() {
        return status == ServiceRequestStatus.RECEIVED || status == ServiceRequestStatus.ASSIGNED;
    }
}

File: frontend/features/reception/types/serviceRequest.ts
export type ServiceRequestStatus = "RECEIVED" | "ASSIGNED" | "CANCELLED";

export interface ServiceRequestListItem {
  id: string;
  requestNumber: string;
  customerName: string;
  customerPhone: string;
  modelName: string | null;
  symptom: string;
  status: ServiceRequestStatus;
  receptionistName: string | null;
  technicianName: string | null;
  visitDate: string | null;
  visitStartTime: string | null;
  visitEndTime: string | null;
  requestedAt: string;
}

export interface ServiceRequestFilters {
  status: ServiceRequestStatus | "";
  sort: "requestedAt,desc" | "requestedAt,asc";
  page: number;
  size: number;
}

export interface ServiceRequestPageData {
  content: ServiceRequestListItem[];
  page: number;
  size: number;
  totalElements: number;
}

export interface CreateServiceRequestRequest {
  customerId: string;
  productId: string | null;
  selectedSubCategoryId: string | null;
  selectedModelName: string | null;
  symptom: string;
  remarks: string;
}

export interface UpdateServiceRequestRequest extends CreateServiceRequestRequest {
  version: number;
}

export interface BaseFeeResponse {
  baseFee: number;
}

export interface ServiceRequestCreateResponse {
  id: string;
  requestNumber: string;
  status: ServiceRequestStatus;
  symptom: string;
  remarks: string;
  baseFee: number;
  requestedAt: string;
}

export interface AvailabilityTimeSlot {
  startTime: string;
  endTime: string;
  available: boolean;
}

export interface AvailableTechnicianResponse {
  technicianId: string;
  technicianName: string;
  slotId: string;
  startTime: string;
  endTime: string;
}

export interface AssignServiceRequestRequest {
  technicianId: string;
  slotId: string;
}

export interface AssignServiceRequestResponse {
  id: string;
  status: ServiceRequestStatus;
  technicianName: string;
  visitDate: string;
  visitStartTime: string;
  visitEndTime: string;
}

export interface ReassignServiceRequestRequest {
    technicianId: string;
    slotId: string;
    version: number;
}


export interface ServiceRequestDetail {
  id: string;
  requestNumber: string;
  customerId: string;
  customerName: string;
  customerPhone: string;
  productId: string | null;
  selectedSubCategoryId: string | null;
  selectedCategoryId: string | null;
  modelName: string | null;
  subCategoryName: string | null;
  symptom: string;
  status: ServiceRequestStatus;
  receptionistName: string | null;
  technicianName: string | null;
  visitDate: string | null;
  visitStartTime: string | null;
  visitEndTime: string | null;
  address: string;
  detailAddress: string;
  remarks: string;
  baseFee: number;
  requestedAt: string;
  assignedAt: string | null;
  cancelledAt: string | null;
  cancelReason: string | null;
  version: number;
}

export interface CancelServiceRequestRequest {
  cancelReason: string;
}

export interface CancelServiceRequestResponse {
  id: string;
  status: ServiceRequestStatus;
  cancelReason: string | null;
  cancelledAt: string;
}